internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,314 @@
+	}
+	// Scaling up.
+	v1beta2conditions.Set(ms, metav1.Condition{
+		Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  clusterv1.MachineSetScalingUpV1Beta2Reason,
+		Message: message,
+	})
+}
+
+func setScalingDownCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool, owner *clusterv1.MachineDeployment) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingDownInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetScalingDownWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	// Deletion is equal to 0 desired replicas.
+	if !ms.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+
+	// Scaling down.
+	if int32(len(machines)) > (desiredReplicas) {
+		messages := []string{fmt.Sprintf("ScalingDown from %d to %d replicas", len(machines), desiredReplicas)}
+
+		machinesStaleDeleting := []string{}
+
+		for _, machine := range machines {
+			if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) >= time.Minute*30 {
+				machinesStaleDeleting = append(machinesStaleDeleting, machine.GetName())
+			}
+		}
+
+		if len(machinesStaleDeleting) > 0 {
+			messages = append(messages, fmt.Sprintf("Machines stuck in deletion for more than 30 minutes: %s", clog.StringListToString(machinesStaleDeleting)))
+		}
+
+		// Scaling is blocked when the bootstrap or infrastructure objects do not exist.
+		if bootstrapObjectNotFound || infrastructureObjectNotFound {
+			messages[0] = fmt.Sprintf("%s is blocked %s", messages[0], calculateMissingReferencesMessage(ms, bootstrapObjectNotFound, infrastructureObjectNotFound))
+		}
+
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionTrue,
+			Reason:  clusterv1.MachineSetScalingDownV1Beta2Reason,
+			Message: strings.Join(messages, ", "),
+		})
+		return
+	}
+
+	// Not scaling down.
+	var message string
+	// The referenced objects are required to exist for the current MachineSet.
+	// Add a message if that's not the case.
+	if isCurrentMachineSet(ms, owner) && (bootstrapObjectNotFound || infrastructureObjectNotFound) {
```

Comment:
Similar to above. I think we can drop this (when we talked about it the last time I wasn't aware of the MachineSet controller changes)